src/tasks/WasmAppBuilder/wasi/WasiAppBuilder.cs

```diff
@@ -0,0 +1,105 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.IO;
+using System.Text.Json.Nodes;
+using Microsoft.Build.Framework;
+
+namespace Microsoft.WebAssembly.Build.Tasks;
+
+public class WasiAppBuilder : WasmAppBuilderBaseTask
+{
+    public bool IsSingleFileBundle { get; set; }
+
+    protected override bool ExecuteInternal()
+    {
+        if (!InvariantGlobalization && string.IsNullOrEmpty(IcuDataFileName))
+            throw new LogAsErrorException("IcuDataFileName property shouldn't be empty if InvariantGlobalization=false");
+
+        if (Assemblies.Length == 0)
+        {
+            Log.LogError("Cannot build Wasm app without any assemblies");
+            return false;
+        }
+
+        var _assemblies = new List<string>();
+        foreach (string asm in Assemblies!)
```

Comment:
Is this running before linking ? Otherwise assemblies don't have to be there in IsSingleFileBundle case.

Reply:
It was unnecessarily copying the assemblies even for `IsSingleFileBundle` case, fixed.